org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright (c) 2014, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.resource.persistence;
+
+import static org.eclipse.xtext.xbase.lib.IterableExtensions.*;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+
+import org.apache.log4j.Logger;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.impl.ExtensibleURIConverterImpl;
+import org.eclipse.xtend.lib.annotations.Accessors;
+import org.eclipse.xtext.generator.AbstractFileSystemAccess2;
+import org.eclipse.xtext.generator.IContextualOutputConfigurationProvider;
+import org.eclipse.xtext.generator.IFileSystemAccessExtension3;
+import org.eclipse.xtext.util.RuntimeIOException;
+import org.eclipse.xtext.xbase.lib.Pure;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+/**
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public class ResourceStorageFacade implements IResourceStorageFacade {
+	private static final Logger LOG = Logger.getLogger(ResourceStorageFacade.class);
+
+	@Inject
+	private IContextualOutputConfigurationProvider outputConfigurationProvider;
+
+	@Inject
+	private Provider<AbstractFileSystemAccess2> fileSystemAccessProvider;
+
+	@Accessors
+	private boolean storeNodeModel = false;
+
+	/**
+	 * @return whether the given resource should be loaded from stored resource state
+	 */
+	@Override
+	public boolean shouldLoadFromStorage(StorageAwareResource resource) {
+		SourceLevelURIsAdapter adapter = SourceLevelURIsAdapter.findInstalledAdapter(resource.getResourceSet());
+		if (adapter == null || adapter.getSourceLevelURIs().contains(resource.getURI()))
+			return false;
+		return doesStorageExist(resource);
+	}
+
+	/**
+	 * Finds or creates a ResourceStorageLoadable for the given resource. Clients should first call
+	 * shouldLoadFromStorage to check whether there exists a storage version of the given resource.
+	 * 
+	 * @return an IResourceStorageLoadable
+	 */
+	@Override
+	public ResourceStorageLoadable getOrCreateResourceStorageLoadable(StorageAwareResource resource) {
+		try {
+			ResourceStorageProviderAdapter stateProvider = head(
```

Comment:
IterableExtensions are still being used here. Can this be simplified, e.g. by using FluentIterable or streams?